include/pybind11/common.h

```diff
@@ -484,6 +484,7 @@ struct internals {
     std::forward_list<void (*) (std::exception_ptr)> registered_exception_translators;
     std::unordered_map<std::string, void *> shared_data; // Custom data to be shared across extensions
     std::vector<PyObject *> loader_patient_stack; // Used by `loader_life_support`
+    std::forward_list<std::string> type_names; // Stores the std::strings holding created types' tp_names
```

Comment:
Minor: might as well name it `static_strings` or the like? (unlike all other fields of `internals`, this one you would never actually read, it's just there for storage with static lifetime and doesn't really have to do much with types)

Also, in case we ever need it again?

```cpp
const char* detail::strdup(const std::string& what) {
    auto& internals = get_internals();
    internals.static_strings.emplace_front(what);
    return internals.static_strings.front().c_str();
}